### PR TITLE
Fix Route 53 pagination when string MaxItems is expected

### PR DESCRIPTION
**Pull request by joguSD**
Route 53 has defined their `MaxItems` pagination field as a string which leads to issues when passing the `page-size` flag to the cli as it will convert the value to an integer. This PR adds a hook to automatically convert the `MaxItems` field back into a string in the case that the operation expects it to be a string.

Fixes: https://github.com/boto/botocore/issues/1063

@JordonPhillips @stealthycoin @kyleknap @jamesls @dstufft

### CODE REVIEW

File: botocore/handlers.py [hunk 1]
```diff
@@ -734,6 +735,17 @@ def convert_body_to_file_like_object(params, **kwargs):
...
 
 
+def fix_route53_max_items(params, model, **kwargs):
+    # Converts the MaxItems param from an integer to a string.
+    # Route53 has defined their MaxItems field for pagination as a string.
+    # This causes issues when the cli automatically converts the page size
+    # to an integer as it will fail validation.
+    if 'MaxItems' in params:
+        if (isinstance(params['MaxItems'], six.integer_types) and
+            isinstance(model.input_shape.members['MaxItems'], StringShape)):
```
Comment (JordonPhillips) [inline botocore/handlers.py]: This should be indented over to not share the same level as the next logical line. If that makes things too long you can also flatten all of it out with guard clauses.

Comment (joguSD) [inline botocore/handlers.py]: I actually researched this yesterday and found PEP8 has no stance. I'll update the indentation to match what exists already.

File: tests/functional/test_route53.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from tests import unittest
+
+import botocore.session
+from botocore.stub import Stubber, ANY
```
Comment (JordonPhillips) [inline tests/functional/test_route53.py]: I don't believe you're using `ANY`

**Comment by kyleknap** [comment]
So one concern that I have with this approach is that while it fixes the the issue with Route53, it may still be an issue for other services that have a string type for their page size. An audit would need to be done to see if there other services do this.

However looking into the botocore code, this is the [line](https://github.com/boto/botocore/blob/develop/botocore/paginate.py#L630) that is causing the issue as we blindly convert the page size into an integer.

Unfortunately, I am not sure if the solution is as easy as removing that line because while we document that page size as an integer, a user could provide a string and it will still work. Thus, removing that line would potentially cause issues.

I think in order to put in a more general solution, we will need to:

1) Remove the line that always converts the page size configuration value to an integer.
2) Include the operation model in the paginator so the types are converted appropriately if on the provided pagination configuration values.

I think I would prefer this approach as it will ensure that any existing or future services do not hit this issue. The only downside with this approach is that the paginators will now have reference to the operation model, but that should not be too big of an issue as the model is already loaded prior to the `get_paginator()` call.

Let me know what you think.

**Comment by joguSD** [comment]
I totally agree that this implementation is not very future proof. I had thought about pushing this fix down to pagination code as you mention but was unsure if the target type would be available at that point. Adding a reference to the shape in the paginators is definitely viable, but would require us to default to int in the case that it's not there, or requiring it and thus having to retroactively update all paginator definitions (and possibly breaking any custom paginators customers may have). I'll look into what's possible.

**Comment by JordonPhillips** [comment]
@joguSD I think he means adding the shapes to the actual `Paginator` python object, not as a new key in the paginator spec.

**Comment by joguSD** [comment]
There actually already is a unit test for this exact situation. It can be seen [here](https://github.com/boto/botocore/blob/develop/tests/unit/test_paginate.py#L722). However, a unit test is insufficient to catch this as it's not an issue with the pagination or validation code independently but both of them combined, hence the functional test.

File: botocore/paginate.py [hunk 3]
```diff
@@ -623,11 +625,17 @@ def _extract_paging_params(self, kwargs):
...
         if page_size is not None:
-            if self._pagination_cfg.get('limit_key', None) is None:
+            if self._limit_key is None:
                 raise PaginationError(
                     message="PageSize parameter is not supported for the "
                             "pagination interface for this operation.")
-            page_size = int(page_size)
+            input_members = self._model.input_shape.members
+            limit_key_shape = input_members.get(self._limit_key)
+            if isinstance(limit_key_shape, StringShape):
```
Comment (joguSD) [inline botocore/paginate.py]: Hmm. I'm not sure how I feel about this. Either way this code will have a dependency on `StringShape`. The way you propose will remove the import but the dependency on `StringShape`'s `type_name` being `'string'` is still there. Can you elaborate on why you want to avoid the import?

**Comment by kyleknap** [comment]
@joguSD I don't think that test applies to what you are trying to fix. The issue you are trying to fix is when `PageSize` is provided in the pagination config as a string as it is modeled as a string in service model. As to a counter argument to just relying on a functional test, I see at least a couple of benefits out of adding the unit tests:

1) If tests start fail, it is easier to determine that it was a change in the pagination code that is causing the issue if there is a unit test for this because a unit test will only start failing if there is an issue with the pagination code. With just a functional test, it is entirely possible that the service definition changed (i.e. changing the type of Route53's `MaxItems` to an integer) and that would cause failures as well, making it a bit more difficult to figure out why it failed

2) If the SDK made a major version bump and for whatever reason the service model definition changed in the major version bump (i.e. making the `MaxItems` an integer), it is entirely possible that the test becomes invalid, we remove it, but then no longer have coverage for this edge case.

In the end, it never really hurts to have the extra bit of testing especially since it should not be that much more effort to add.

File: botocore/paginate.py (hunk 3 above)
Comment (kyleknap) [inline botocore/paginate.py]: I prefer `type_name` over `StringShape` for the following reasons:

1) It is more consistent throughout the botocore codebase. Just grepping through the botocore codebase `type_name` is used considerably more than `StringShape`. When grepping for `StringShape`, I see it only being used a couple of times in the `model` module

2) It is entirely possible for us to change the class name of `StringShape`, but it is way less possible for the `type_name` to ever change. This value is directly pulled from the JSON model and if the literal value `"string"` was to ever change it would break not just our SDK, but probably all of the other SDKs.

3) It seems weird that we would import `StringShape` just for an `isinstance` check when we can just use `type_name`. Relying on an import for instance checks seems to defeat the purpose of having all of the shape classes from the `model` module subclassing from the `Shape` class and having a `type_name` property.

Comment (joguSD) [inline botocore/paginate.py]: Good points, I'll change this over.
